### ISSUE TITLE
RFC: add support for creating a new array by tiling a provided array

### ISSUE THREAD

**Issue by kgryte** [data-apis/array-api#655]
This RFC proposes adding support to the array API specification for creating a new array by tiling a provided array.

## Overview

Based on array comparison [data](https://github.com/data-apis/array-api-comparison/blob/fa017050ea8079c0fbca1142dcfd130bdcbb6136/signatures/manipulation/tile.md), the API is available in all array libraries.

## Prior art

- NumPy: <https://numpy.org/doc/stable/reference/generated/numpy.tile.html>
- PyTorch: <https://pytorch.org/docs/stable/generated/torch.tile.html>
- TensorFlow: <https://www.tensorflow.org/api_docs/python/tf/tile>
- JAX: <https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.tile.html>
- CuPy: <https://docs.cupy.dev/en/stable/reference/generated/cupy.tile.html>
- Dask: <https://docs.dask.org/en/stable/generated/dask.array.tile.html>

## Proposal

```python
def tile(x: array, repetitions: Union[int, Tuple[int, ...]], /)
```

- **repetitions**: the number of repetitions of `x` along each axis. Let `N = len(repetitions)` and let `M = len(x.shape)`.

  - if `M < N`, then `x` must be broadcasted to be `N` dimensional by prepending new axes.
  - if `M > N`, then `repetitions` must be broadcasted by prepending new axes.

## Questions

- Neither PyTorch nor TensorFlow support providing an `int` for `repetitions`. TensorFlow requires `repetitions` to be a `Tensor`. All array libraries apart from TensorFlow allow `repetitions` to be a tuple. Should `repetitions` be typed more strictly?

## Related

- Mention of xarray's need for `tile`: <https://github.com/data-apis/array-api/issues/187#issuecomment-1553615779>

**Comment by rgommers** [comment]
> * Neither PyTorch nor TensorFlow support providing an `int` for `repetitions`. TensorFlow requires `repetitions` to be a `Tensor`. All array libraries apart from TensorFlow allow `repetitions` to be a tuple. Should `repetitions` be typed more strictly?

I'd support tuple-only. It can always be loosened in the future, and this function isn't used all that much hence it's not much of a burden to type `(3,)` instead of `3`.

Should `repetitions` be positional-only perhaps? I noticed that NumPy uses `reps` rather than `repetitions`.

**Comment by kgryte** [comment]
> Should repetitions be positional-only perhaps? I noticed that NumPy uses reps rather than repetitions.

The current proposal has `repetitions` as positional-only. Hence, the `reps`/`repetitions` distinction should not matter.

The question then becomes should `repetitions` be positional or keyword? In which case, settling on a naming convention makes sense. For this, there is divergence in the ecosystem (e.g., PyTorch uses `dims`, TensorFlow uses `multiples`).

**Comment by kgryte** [comment]
Regarding usage, according to usage data for [common APIs](https://github.com/data-apis/array-api-comparison/blob/fa017050ea8079c0fbca1142dcfd130bdcbb6136/data/common_apis_ranks.csv#L50), `tile` does seem to be used a fair amount (e.g., more than various element-wise, linalg, and other APIs already included in the specification).
